src/ledger/recovery.zig

```diff
@@ -0,0 +1,500 @@
+        // Existing shreds do not need to be freed because they were already owned by calling scope.
+        for (all_shreds, shreds.mask, 0..num_recovered_so_far) |shred, was_present, _| {
+            if (!was_present) shred.deinit();
+        }
+    }
+    std.debug.assert(shreds.input_shreds.len == shreds.shards.len);
+    for (shreds.input_shreds, shreds.shards, 0..) |maybe_shred, maybe_shard, index| {
+        if (maybe_shred) |shred| {
+            all_shreds[index] = shred;
+        } else {
+            const shard = maybe_shard orelse return error.TooFewShards;
+            all_shreds[index] = try reconstructShred(allocator, meta, shard, index);
+        }
+        num_recovered_so_far += 1;
+    }
+
+    try setMerkleProofs(allocator, meta, shreds, all_shreds);
+
+    return all_shreds;
+}
+
+fn reconstructShred(
+    allocator: Allocator,
+    meta: RecoveryMetadata,
+    shard: []const u8,
+    index: usize,
+) !Shred {
+    if (index < meta.code_header.num_data_shreds) {
+        const data_shred = try DataShred.fromRecoveredShard(
+            allocator,
+            meta.common_header.signature,
+            meta.chained_merkle_root,
+            meta.retransmitter_signature,
+            shard,
+        );
+        const this = data_shred.fields.common;
+        const set = meta.common_header;
+        if (this.variant.proof_size != set.variant.proof_size or
+            this.variant.chained != set.variant.chained or
+            this.variant.resigned != set.variant.resigned or
+            this.slot != set.slot or
+            this.version != set.version or
+            this.fec_set_index != set.fec_set_index)
+        {
+            return error.InvalidRecoveredShred;
+        }
+        return .{ .data = data_shred };
+    } else {
+        const offset = index - meta.code_header.num_data_shreds;
+        var this_common_header = meta.common_header;
+        var this_code_header = meta.code_header;
+        this_common_header.index += @intCast(offset);
+        this_code_header.position = @intCast(offset);
+        const code_shred = try CodeShred.fromRecoveredShard(
+            allocator,
+            this_common_header,
+            this_code_header,
+            meta.chained_merkle_root,
+            meta.retransmitter_signature,
+            shard,
+        );
+        return .{ .code = code_shred };
+    }
+}
+
+fn setMerkleProofs(
+    allocator: Allocator,
+    meta: RecoveryMetadata,
+    shreds: RecoveryShreds,
+    all_shreds: []Shred,
+) !void {
+    // Compute merkle tree
+    var tree = try std.ArrayList(Hash).initCapacity(allocator, all_shreds.len);
+    defer tree.deinit();
+    for (all_shreds) |shred| {
+        tree.appendAssumeCapacity(try shred.merkleNode());
+    }
+    try makeMerkleTree(&tree);
+
+    // set the merkle proof on the recovered shreds.
+    const num_shards: usize =
+        @intCast(meta.code_header.num_data_shreds + meta.code_header.num_code_shreds);
```

Comment:
Don't think you need this `@intCast`, `num_data_shreds` and `num_code_shreds` are `u16`s.

Reply:
9d24462d596fb2aae458c3483045e2240155f468